parser.y

```diff
@@ -331,19 +311,219 @@ primary_expr: ID { $$ = std::make_unique<IdExprNode>(Loc(@1), $1); }
+      param_types.push_back(param->type->Clone());
+    }
+    auto type = std::make_unique<FuncType>(/* return */ std::move(decl->type), std::move(param_types));
+    $$ = std::make_unique<FuncDefNode>(Loc(@1), std::move(decl->id), std::move(params), nullptr, std::move(type));
+  }
+  /* TODO: identifier list */
+  /* The identifier may be a type name. */
+  /* TODO: direct declarator ( identifier list ) */
   ;
 
-pointer_type: type_specifier STAR { $$ = std::make_unique<PtrType>($1); }
+pointer_opt: pointer { $$ = $1; }
+  | epsilon { $$ = 0; }
   ;
 
+/* A pointer is a sequence of one or more '*'s. */
+pointer: STAR { $$ = 1; }
+  | pointer STAR { $$ = $1 + 1; }
+  ;
+
+parameter_type_list_opt: parameter_type_list { $$ = $1; }
+  | epsilon { $$ = std::vector<std::unique_ptr<ParamNode>>{}; }
+  ;
+
+parameter_type_list: parameter_list { $$ = $1; }
+  /* TODO: parameter list, ... */
+  ;
+
+parameter_list: parameter_declaration {
+    $$ = std::vector<std::unique_ptr<ParamNode>>{};
+    $$.push_back($1);
+  }
+  | parameter_list COMMA parameter_declaration {
+    auto parameter_list = $1;
+    parameter_list.push_back($3);
+    $$ = std::move(parameter_list);
+  }
+  ;
+
+parameter_declaration: declaration_specifiers declarator {
+    auto decl = $2;
+    auto resolved_type = ResolveTypeAs(std::move(decl->type), $1);
+    $$ = std::make_unique<ParamNode>(Loc(@2), std::move(decl->id), std::move(resolved_type));
+  }
+  /* Declare parameters without identifiers. */
+  | declaration_specifiers abstract_declarator_opt {
+    // TODO: for type names
+    $$ = nullptr;
+  }
+  ;
+
+abstract_declarator_opt: abstract_declarator
+  | epsilon
+  ;
+
+/* 6.7.6 Type names */
+/* NOTE: abstract means the declarator does not have an identifier */
+abstract_declarator: pointer
+  | pointer_opt direct_abstract_declarator
+  ;
+
+direct_abstract_declarator: LEFT_PAREN abstract_declarator RIGHT_PAREN
+  | direct_abstract_declarator_opt LEFT_SQUARE NUM RIGHT_SQUARE
+  | direct_abstract_declarator_opt LEFT_PAREN parameter_type_list_opt RIGHT_PAREN
+  ;
+
+direct_abstract_declarator_opt: direct_abstract_declarator
+  | epsilon
+  ;
+
+/* 6.7.8 Initialization */
+/* The current object shall have array type and the expression shall be an integer constant expression. */
+initializer: LEFT_CURLY initializer_list comma_opt RIGHT_CURLY { $$ = $2; }
+  | expr { $$ = $1; }
+  ;
+
+/* TODO: the initializer may be nested */
+initializer_list: expr {
+    $$ = std::vector<std::unique_ptr<ExprNode>>{};
+    $$.push_back($1);
+  }
+  | initializer_list COMMA expr {
+    auto initializer_list = $1;
+    initializer_list.push_back($3);
+    $$ = std::move(initializer_list);
+  }
+  ;
+
+comma_opt: COMMA
+  | epsilon
+  ;
 
 epsilon: %empty;
 %%
 
 void yy::parser::error(const yy::location& loc, const std::string& err) {
   std::cerr << loc << ": " << err << std::endl;
 }
+
+namespace {
+
+std::unique_ptr<Type> ResolveTypeAs(std::unique_ptr<Type> type,
```

Comment:
Can you give a simple example for this function?

Reply:
Let's take a complex declaration as an example:
```c
int* ptr_arr[10];
```

- The declaration is recognized by the rule `decl: declaration_specifiers declarator SEMICOLON` (omit `init` for simplicity).
  - `int` is recognized by the non-terminal `declaration_specifier`, which constructs _an int type_ in its action.
  - `* ptr_arr[10]` is recognized by the rule `declarator: pointer_opt direct_declarator`:
    - The `pointer_opt` has its semantic value as `1`.
    - The `direct_declarator` has its semantic value as _an array to an unknown type_.
    - So in the following semantic action, we try to combine the two types into a single type: _an array to a pointer to an unknown type_.
      ```yacc
      declarator: pointer_opt direct_declarator {
          @$ = @2; // Set the location to the identifier.
          auto declarator = $2;
          for (int i = 0, e = $1; i < e; ++i) {
            auto unknown_ptr_type = std::make_unique<PtrType>(std::make_unique<PrimType>(PrimitiveType::kUnknown));
            declarator->type = ResolveTypeAs(std::move(declarator->type), std::move(unknown_ptr_type));
          }
          $$ = std::move(declarator);
        }
        ;
      ```
      To achieve this, we have to place the type _a pointer to an unknown type_ as the innermost type of _an array to a pointer to an unknown type_, which is what the function `ResolveTypeAs` does.
  - So in the following semantic action, we try to combine the type _an int type_ with the type _an array to a pointer to an unknown type_:
    ```yacc
    decl: declaration_specifiers init_declarator SEMICOLON {
        auto decl = $2;
        decl->type = ResolveTypeAs(std::move(decl->type), $1);
        $$ = std::move(decl);
      }
      ;
    ```
    This is achieved by placing _an int type_ as the innermost type of _an array to a pointer to an unknown type_.

The `ResolveTypeAs` function is implemented as a recursive function. It gets directly to the innermost type, using the `resolved_type` as the innermost type and wraps it with the layers of `type`. Take the above _an array to a pointer to an unknown type_ resolution as an example:
- The outermost type is an array, thus the following condition is entered:
    ```cpp
      if (type->IsArr()) {
        auto arr_type = static_cast<ArrType*>(type.get());
        auto type_ = ResolveTypeAs(arr_type->element_type().Clone(), std::move(resolved_type));
        return std::make_unique<ArrType>(std::move(type_), arr_type->len());
      }
    ```
    It recursively tries to resolve its element type _a pointer to an unknown type_.
    - The type is now a pointer, thus the following condition is entered:
      ```cpp
        if (type->IsPtr()) {
          auto ptr_type = static_cast<PtrType*>(type.get());
          auto type_ = ResolveTypeAs(ptr_type->base_type().Clone(), std::move(resolved_type));
          return std::make_unique<PtrType>(std::move(type_));
        }
      ```
      It tries to resolve its base type, which is _an unknown type_.
      - Finally, the type is a primitive type _an unknown type_. The base condition is entered:
        ```cpp
        if (type->IsPrim()) {
          return resolved_type;
        }
        ```
        The `resolved_type` is the _an int type_ in this case.
- Now, we go all the way up.
  - `return std::make_unique<PtrType>(std::move(type_ /* an int type */));`
  - `return std::make_unique<ArryType>(std::move(type_ /* a pointer to an int type */));`
  Getting the resolved type _an array to a pointer to an int type_.

I noticed that in the assignment `auto type_ = ResolveTypeAs(arr_type->element_type().Clone(), std::move(resolved_type));`, the variable `type_` should better be named as `resolved_type`, as the type it holds is resolved. How do you think about this? The parameter `resolved_type` can be reassigned in this case. The following is the concept code:
```cpp
  if (type->IsPtr()) {
    auto ptr_type = static_cast<PtrType*>(type.get());
    resolved_type = ResolveTypeAs(ptr_type->base_type().Clone(), std::move(resolved_type));
    return std::make_unique<PtrType>(std::move(resolved_type));
  }
```

Reply:
> How do you think about this? The parameter resolved_type can be reassigned in this case. The following is the concept code:

Sure! I second this.